src/Shapes/ShapeFactory.ts

```diff
@@ -0,0 +1,112 @@
+    const styleInfo = {
+      color: info.parameters.color || drawState.getAlphaColor(),
+      thickness: info.parameters.thickness || drawState.thickness,
+    };
+    const shape = this.create(info.type, styleInfo);
+    // info.parameters is the basePoint when a shape is created by the drawer
+    if (info.parameters instanceof Point) {
+      this.setOriginPointshape(shape, info.parameters);
+    } else {
+      this.setInfo(shape, info.parameters);
+    }
+    return shape;
+  }
+
+  create(shapeType : ShapeType, styleInfo : { color: AlphaColor, thickness : number }) : Shape {
+    switch (shapeType) {
+      case ShapeType.Draw:
+        return new Draw(styleInfo.color, styleInfo.thickness);
+      case ShapeType.RectangleFull:
+        return new RectangleFull(styleInfo.color, styleInfo.thickness);
+      case ShapeType.RectangleStroke:
+        return new RectangleStroke(styleInfo.color, styleInfo.thickness);
+      case ShapeType.EllipseFull:
+        return new EllipseFull(styleInfo.color, styleInfo.thickness);
+      case ShapeType.EllipseStroke:
+        return new EllipseStroke(styleInfo.color, styleInfo.thickness);
+      case ShapeType.Line:
+        return new Line(styleInfo.color, styleInfo.thickness);
+      case ShapeType.Fill:
+        return new Fill(styleInfo.color, styleInfo.thickness);
+      default:
+        throw new Error('Shape is not implemented');
+    }
+  }
+
+  setInfo(shape : Shape, parameter : any) : void {
+    if (shape instanceof UpdatableShape) {
+      shape.startDate = parameter.startDate as number;
+      shape.endDate = parameter.endDate as number;
+    }
+
+    if (shape instanceof Fill || shape instanceof DraggableShape) {
+      this.setOriginPointshape(shape, parameter.originPoint as Point);
+    }
+
+    if (shape instanceof DraggableShape) {
+      shape.height = parameter.height as number;
+      shape.width = parameter.width as number;
+    }
+
+    if (shape instanceof Draw) {
+      shape.points = parameter.points as Array<Point>;
+    }
+  }
+
+  setOriginPointshape(shape : Shape, basePoint : Point) : void {
+    if (shape instanceof Fill || shape instanceof DraggableShape) {
+      shape.originPoint = basePoint;
+    } else if (shape instanceof Draw) {
+      shape.points.push(basePoint);
+    }
+  }
+
+  serialize(elem: Shape): IShapeInfo {
+    return { type: this.getType(elem), parameters: elem.exportInfo() };
+  }
+
+  getType(shape: Shape) : ShapeType {
+    switch (shape.constructor) {
+      case Draw:
+        return ShapeType.Draw;
+      case RectangleFull:
+        return ShapeType.RectangleFull;
+      case RectangleStroke:
+        return ShapeType.RectangleStroke;
+      case EllipseFull:
+        return ShapeType.EllipseFull;
+      case EllipseStroke:
+        return ShapeType.EllipseStroke;
+      case Line:
+        return ShapeType.Line;
+      case Fill:
+        return ShapeType.Fill;
+      default:
+        throw new Error('Shape is not implemented');
+    }
+  }
+}
```

Comment:
Oui je réflechis. De base je voulais faire un dictionnaire<type, shapeType> comme en C# qui associe à chaque Shape son type. Mais j'ai pas réussi.
Je retente ma chance ou je fais ta solution 👍

Reply:
J'ai réussi un faire un truc comme ca: 
```ts
  shapesDictionnary : Array<{ proto: Shape, shapeType: ShapeType }> = [
    { proto: Draw.prototype, shapeType: ShapeType.Draw },
    { proto: RectangleFull.prototype, shapeType: ShapeType.RectangleFull },
    { proto: RectangleStroke.prototype, shapeType: ShapeType.RectangleStroke },
    { proto: EllipseFull.prototype, shapeType: ShapeType.EllipseFull },
    { proto: EllipseStroke.prototype, shapeType: ShapeType.EllipseStroke },
    { proto: Line.prototype, shapeType: ShapeType.Line },
    { proto: Fill.prototype, shapeType: ShapeType.Fill },
  ];
```
 mais j'ai des erreurs obscure lors de l'instanciation :
```ts
create(shapeType : ShapeType, styleInfo : { color: AlphaColor, thickness : number }) : Shape {
    const shapeInfo = this.shapesDictionnary.find((shp) => shp.shapeType === shapeType);
    if (shapeInfo === undefined) {
      throw new Error(`Shape ${shapeType.toString()} is not implemented`);
    }
    
    const shape = Object.create(shapeInfo.proto);
    const instance = shape.constructor.apply(shape, [styleInfo.color, styleInfo.thickness]);
    return instance as Shape;
  }
```
Je regarde ca demain 🤔

Reply:
So after a bit of thinking and reading, I've noticed this:

- `ShapeFactory.getType` is currently being used only in `ShapeFactory.serialize`.
- `ShapeFactory.serialize` is currently being used only in `ShapeManager.emit`.

Now if we take a look at `ShapeFactory.getType` function:
```ts
getType(shape: Shape) : ShapeType {
    switch (shape.constructor) {
      case Draw:
        return ShapeType.Draw;
      case RectangleFull:
        return ShapeType.RectangleFull;
      // ...
      default:
        throw new Error('Shape is not implemented');
    }
}
  ```

We are intaking a shape, and the only actual code in here is the check in switch, that uses the shape parameter: `switch (shape.constructor) {`. To me, this screams Feature Envy [G14]. This function would be much better in Shape class, then we could do shape.getType();

Now let's take a look at `ShapeFactory.serialize` function:
```ts
serialize(elem: Shape): IShapeInfo {
    return { type: this.getType(elem), parameters: elem.exportInfo() };
}
```

The function intakes a Shape parameter, and uses its properties to do something: this reeks of Feature Envy [G14] too. We should move serialize function inside Shape class, so as to do elem.serialize().

The only place where serialize is used is:
```ts
  emit() : void {
    if (this.currentShape === null) return;
    this.notify({
      type: ActionType.shape,
      parameters: this.factory.serialize(this.currentShape),
    });
  }

```
Which could then be refactored into:
```ts
  emit() : void {
    if (this.currentShape === null) return;
    this.notify({
      type: ActionType.shape,
      parameters: this.currentShape.serialize(),
    });
  }
```

Reply:
Now, regarding the actual implementation of getType, as we've seen, it is only used in serialize, and serialize is only used in emit.

> De base je voulais faire un dictionnaire<type, shapeType>

The problem I have with a Dictionary is that we would have to actually remember to update it whenever we create a shape; appart from the Error you're throwing in the create function, we can't be sure that each Shape actually has a line in the dictionary. The Error would only be thrown when we try to draw this very shape, now this would mean that it could easily slip away.

Now, a shapeType property would be ideal regarding this case: if we don't actually create a shapeType property in our subclass, typescript will scream at us from the get-go.

> L'avantage par rapport à ta solution est que ca permet d'instancier une shape depuis sont ShapeType.

So while this is true, I am not certain whether it is necessary. This feels like it breaks the [YAGNI Principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), but on the other hand, it sure would be a great idea.

So here's what I would do, to compromise between the two:

1. As discussed, move both getType and serialize function inside of Shape class, as public method.
2. As of right now, create a **private** shapeType: ShapeType attribute: this way, we enforce the shapeType property in every single one of our `Shape`s.
3. The property being private, we can only access it through the getType function, which would now actually be a getter:
```ts
// Shape.ts
getType() : ShapeType {
   return this.shapeType;
}
```
4. Now, if and when we need to have a Dictionary, we could simply update the getType function to use the Dictionary and remove the property from the Shape class.

This feels like the ideal solution; it does allow for improvement in case we ever need a dictionary, because we would only the getType function to access the type!

Reply:
Done cf commit. Ca marche nikel et ca fait le taff.
Je te laisse resolve si tu valides

Sur le plan purement architectural je me pose des questions sur la pattern le plus optimal. Le pattern Resolver / Factory ([cf un exemple que j'aime bien)](https://github.com/HandyOrg/HandyControl/blob/master/src/Shared/HandyControl_Shared/Controls/PropertyGrid/PropertyResolver.cs) est super propre et permet de déléguer des responsabilités. La on fout un peu tout dans shape nottament la responsabilité de se serialiser... 🤔  
Après ts est pas aussi avancer qu'un bon Java / C# et ca peut être compliqué d'implémenter des patterns trés abstraits.